### PR TITLE
Fix #2811: Adapt to separated parallel collections in 2.13

### LINKED ISSUE #2811: Adjust to Scala 2.13 moving parallel collections into a module
we are interested in keeping Scala.js in the Scala 2.13 community build continuously throughout the 2.13 cycle, if at all possible, because Scala.js is so important in itself, and because so many other projects in the community build depend on it

after the merge of https://github.com/scala/scala/pull/5603, Scala.js no longer compiles on 2.13 because of the references to parallel collections in scalalib/overrides-2.13/scala/collection/immutable/Range.scala (see [example failure log](https://scala-ci.typesafe.com/view/scala-2.13.x/job/scala-2.13.x-integrate-community-build/283/consoleFull))

I thought I might try to patch it myself, but I don't really understand what the `overrides` directories are for, and when I diff your Range.scala against the one in Scala 2.13 HEAD, I see other differences besides just ones related to parallel collections.

so... is this a change y'all could do yourselves...? 🙏

### PR DESCRIPTION
**Pull request by gzm0**
Parallel collections got moved to a separate library.

### CODE REVIEW

**Comment by gzm0** [comment]
Review by @sjrd and @SethTisue. I have manually verified that this compiles again with:

```
set resolvers in compiler += "scala-integration" at "https://scala-ci.typesafe.com/artifactory/scala-integration/"
set resolvers in scalalib += "scala-integration" at "https://scala-ci.typesafe.com/artifactory/scala-integration/"
++2.13.0-pre-e2a2cba
scalalib/compile
```

**Comment by sjrd** [comment]
The above changes look good, but what about:

* [`ParIncOptimizer.scala`](https://github.com/scala-js/scala-js/blob/master/tools/jvm/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/ParIncOptimizer.scala#L14) (test with `tools/compile`)
* [`RhinoLinkFailureTest.scala`](https://github.com/scala-js/scala-js/blob/master/no-ir-check-test/src/test/scala/org/scalajs/testsuite/noircheck/RhinoLinkFailureTest.scala) (test with `noIRCheckTest/test:compile`)

?

(found with `git grep 'collection.parallel'`)

**Comment by sjrd** [comment]
Btw in our build I do the following to add Scala resolvers *everywhere*:

    set resolvers in Global += "scala-integration" at "https://scala-ci.typesafe.com/artifactory/scala-integration/"

(unless it's `in ThisBuild`? I think both work in this case)

**Comment by gzm0** [comment]
Hmmm... I can't pull `scala-parallel-collections` from `scala-integration` :-/

@SethTisue has it been published? Our linker actually relies on parallel collections.

```
[info] Resolving org.scala-lang#scala-parallel-collections_2.13.0-pre-e2a2cba;1.0.0-SNAPSHOT ...
[warn] 	module not found: org.scala-lang#scala-parallel-collections_2.13.0-pre-e2a2cba;1.0.0-SNAPSHOT
[warn] ==== local: tried
[warn]   /Users/tschlatt/.ivy2/local/org.scala-lang/scala-parallel-collections_2.13.0-pre-e2a2cba/1.0.0-SNAPSHOT/ivys/ivy.xml
[warn] ==== public: tried
[warn]   https://repo1.maven.org/maven2/org/scala-lang/scala-parallel-collections_2.13.0-pre-e2a2cba/1.0.0-SNAPSHOT/scala-parallel-collections_2.13.0-pre-e2a2cba-1.0.0-SNAPSHOT.pom
[warn] ==== scala-integration: tried
[warn]   https://scala-ci.typesafe.com/artifactory/scala-integration/org/scala-lang/scala-parallel-collections_2.13.0-pre-e2a2cba/1.0.0-SNAPSHOT/scala-parallel-collections_2.13.0-pre-e2a2cba-1.0.0-SNAPSHOT.pom
[info] Resolving jline#jline;2.14.3 ...
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	::          UNRESOLVED DEPENDENCIES         ::
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	:: org.scala-lang#scala-parallel-collections_2.13.0-pre-e2a2cba;1.0.0-SNAPSHOT: not found
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn] 	Note: Unresolved dependencies path:
[warn] 		org.scala-lang:scala-parallel-collections_2.13.0-pre-e2a2cba:1.0.0-SNAPSHOT (/Users/tschlatt/20p/scala-js/project/Build.scala#L582)
[warn] 		  +- org.scala-js:scalajs-tools_2.13.0-pre-e2a2cba:0.6.15-SNAPSHOT ()
```

**Comment by SethTisue** [comment]
I'll work on getting a 0.1.0 release published and let you know, here, when it's ready

**Comment by SethTisue** [comment]
@gzm0 here ya go https://repo1.maven.org/maven2/org/scala-lang/modules/scala-parallel-collections_2.13.0-pre-e2a2cba/0.1.1/

you may hit https://github.com/scala/scala-parallel-collections/issues/22 though...

**Comment by SethTisue** [comment]
until we have a solution for that, I guess you could put your own empty `scala.collection.parallel.CollectionConverters` class in a `src/scala-2.13` directory of your own? thinking out loud...

**Comment by sjrd** [comment]
Here is one way with which we can write cross-compilable code: https://github.com/scala/scala-parallel-collections/issues/22#issuecomment-288389306

**Comment by gzm0** [comment]
Potentially blocked by scala/scala-parallel-collections#23. Investigating if `toParArray` is really necessary.

**Comment by gzm0** [comment]
Updated. Tools compile as well now.

**Comment by gzm0** [comment]
QQ: Should this target the 0.6 branch? I'm assuming the community build runs against the master branch. If so, we should probably re-attach the issue. It's currently on the 0.6.16 milestone.

**Comment by sjrd** [comment]
This PR should definitely target the 0.6.x branch. Otherwise if 2.13.0-M1 is released before Scala.js 1.0.0 (which I guess has a probability sufficiently high that we should cater for this), we won't be able to release Scala.js 0.6.y for 2.13.0-M1.

Now, that's pretty orthogonal to whether the community build should monitor master or 0.6.x. Let's discuss this somewhere else: https://github.com/scala/community-builds/issues/506

**Comment by gzm0** [comment]
Good point. Rebasing and fixing the Rhino error as well :)

File: project/Build.scala [hunk 1]
```diff
@@ -588,6 +588,11 @@ object Build {
               "org.scala-js" % "closure-compiler-java-6" % "v20160517",
               "com.googlecode.json-simple" % "json-simple" % "1.1.1" exclude("junit", "junit"),
               "com.novocode" % "junit-interface" % "0.9" % "test"
+          ) ++ (
+              if (scalaVersion.value.startsWith("2.13"))
```
Comment (SethTisue) [inline project/Build.scala]: I'd suggest:

```
CrossVersion.partialVersion(scalaVersion.value) match {
  case Some((2, n)) if n >= 13 => ...
  case _ => ...
}
```

**Comment by gzm0** [comment]
@sjrd I'm unable to fix the Rhino linking test with a missing JDK thing. And for 2.13 we won't have anything on the classpath that (trivially) doesn't link. Are you fine if I just drop the test? It's not like we are going to invest in Rhino anyways. The alternative is to introduce build foo to get something on the compile classpath which is not on the link classpath.

**Comment by gzm0** [comment]
Ok. So the problem simply is that Rhino does *not* throw an exception when it fails to load a class. The reason the test passed is that the *call* to the module accessor (obviously) type errors:

```
TypeError: Cannot find function sc_parallel_Splitter$ in object [object Object].
```

This passed our rudimentary test:

```scala
assertTrue(e.toString.contains("sc_parallel_Splitter$"))
```

But of course wouldn't behind any sort of renaming. I propose to simply:

- File this as a wontfix bug.
- Remove the test.

**Comment by gzm0** [comment]
Updated

**Comment by gzm0** [comment]
@sjrd IIRC, you know / have the necessary permissions to start a community build on this one. Could you do so? :)

**Comment by sjrd** [comment]
@SethTisue I have prepared in community-builds branch to test the fix in this PR at https://github.com/sjrd/community-builds/tree/test-scalajs-parallel-colls
Could you schedule a community build off that branch, please?

**Comment by sjrd** [comment]
```
[info] Scala.js tools: found 6 potential binary incompatibilities while checking against org.scala-js:scalajs-tools_2.10:0.6.15 
[error]  * method apply(scala.collection.immutable.List)java.util.concurrent.atomic.AtomicReference in object org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils#AtomicAcc's type is different in current version, where it is (scala.collection.parallel.immutable.ParVector)java.util.concurrent.atomic.AtomicReference instead of (scala.collection.immutable.List)java.util.concurrent.atomic.AtomicReference
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils#AtomicAcc.apply")
[error]  * method prepAdd(scala.collection.parallel.mutable.ParArray)java.util.concurrent.atomic.AtomicReference in object org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer#CollOps in current version does not have a correspondent with same parameter signature among (scala.collection.GenIterableLike)java.lang.Object, (scala.collection.parallel.immutable.ParVector)java.util.concurrent.atomic.AtomicReference
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer#CollOps.prepAdd")
[error]  * method finishAdd(java.util.concurrent.atomic.AtomicReference)scala.collection.parallel.mutable.ParArray in object org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer#CollOps has a different result type in current version, where it is scala.collection.parallel.immutable.ParVector rather than scala.collection.parallel.mutable.ParArray
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer#CollOps.finishAdd")
[error]  * method emptyParIterable()scala.collection.parallel.mutable.ParArray in object org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer#CollOps has a different result type in current version, where it is scala.collection.parallel.immutable.ParVector rather than scala.collection.parallel.mutable.ParArray
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer#CollOps.emptyParIterable")
[error]  * extension method removeAll$extension(java.util.concurrent.atomic.AtomicReference)scala.collection.immutable.List in object org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils#AtomicAccOps has a different result type in current version, where it is scala.collection.parallel.immutable.ParVector rather than scala.collection.immutable.List
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils#AtomicAccOps.removeAll$extension")
[error]  * method removeAll()scala.collection.immutable.List in class org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils#AtomicAccOps has a different result type in current version, where it is scala.collection.parallel.immutable.ParVector rather than scala.collection.immutable.List
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.scalajs.core.tools.linker.frontend.optimizer.ConcurrencyUtils#AtomicAccOps.removeAll")
java.lang.RuntimeException: Scala.js tools: Binary compatibility check failed!
```

**Comment by gzm0** [comment]
Ah yes, of course :) When targeting the old stuff :P

**Comment by SethTisue** [comment]
community build run: https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/1379/consoleFull (404 til worker spins up)

**Comment by gzm0** [comment]
Updated (not tested locally due to time constraints).

**Comment by gzm0** [comment]
Updated

**Comment by sjrd** [comment]
CB says

    Project scala-js----------------: SUCCESS (project rebuilt ok)

**Pull request merged**